dali/pipeline/operators/decoder/decoder_cache.h

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_DECODER_CACHE_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_DECODER_CACHE_H_
+
+#include <string>
+#include <unordered_map>
+#include "dali/pipeline/data/tensor_list.h"  // needed for Dims
+#include "dali/kernels/span.h"
+#include "dali/error_handling.h"
+
+namespace dali {
+
+class DecoderCache {
+ public:
+    using ImageKey = std::string;
+    struct DecodedImage {
+        span<uint8_t, dynamic_extent> data;
+        Dims dims;
+
+        inline bool operator==(const DecodedImage& oth) const {
+            return data == oth.data
+                && dims == oth.dims;
+        }
+    };
+
+    explicit DecoderCache(std::size_t cache_size = (1<<26))
+        : cache_size_(cache_size) {
+        buffer_ = static_cast<uint8_t*>(GPUBackend::New(cache_size_, false));
+        DALI_ENFORCE(buffer_ != nullptr);
+        tail_ = buffer_;
+        buffer_end_ = buffer_ + cache_size;
+    }
+
+    ~DecoderCache() {
+        if (buffer_)
+            GPUBackend::Delete(buffer_, 0, false);
+    }
+
+    DISABLE_COPY_MOVE_ASSIGN(DecoderCache);
+
+    inline bool IsCached(const ImageKey& image_key) const {
+        return (cache_.find(image_key) != cache_.end());
+    }
+
+    inline const Dims& GetShape(const ImageKey& image_key) const {
+        DALI_ENFORCE(!image_key.empty());
+        const auto it = cache_.find(image_key);
+        DALI_ENFORCE(it != cache_.end(),
+            "cache entry " + image_key + " not found");
+        return it->second.dims;
+    }
+
+    inline void CopyData(const ImageKey& image_key,
+                         void* destination_buffer) const {
+        std::cout << "CopyData: image_key[" << image_key << "]" << std::endl;
+        DALI_ENFORCE(!image_key.empty());
+        DALI_ENFORCE(destination_buffer != nullptr);
+        const auto it = cache_.find(image_key);
+        DALI_ENFORCE(it != cache_.end(),
+            "cache entry " + image_key + " not found");
+        const auto &data = it->second.data;
+        MemCopy(destination_buffer,
+                data.data(),
+                data.size(),
+                0);  // TODO(janton): provide cuda stream
```

Comment:
You probably need to provide this stream as an argument and stream will be known by nvJPEG operator internally (it uses couple of them in parallel).